File: src/learnkotlin/advanced/reflection/TypeErasure.java
package learnkotlin.advanced.reflection;

import java.util.ArrayList;
import java.util.List;

public class TypeErasure {

    public static void main(String[] args) {
        List<String> strings = new ArrayList<>();
        List<Integer> integers = new ArrayList<>();

        printlnList(strings);
        printlnList(integers);

    }

    private static <T> void printlnList(List<T> list) {

    }
}

File: src/learnkotlin/advanced/reflection/TypeErasure.kt
package learnkotlin.advanced.reflection

fun main() {
    val strings = arrayListOf("one", "two", "three")
    val integers = arrayListOf(1, 3, 5)


    printObject("Hello")
    printObject(strings)
    printObject(setOf<Set<String>>())
}

fun printList(list: List<*>) {
    println(list)
}

fun <T> printObject(obj: T) = when (obj) {
    is Int -> println("this is int")
    is String -> println("this is string")
    is List<*> -> println(obj.joinToString())
    else -> println("undefined")
}